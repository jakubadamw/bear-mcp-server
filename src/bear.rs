use crate::Result;

use serde::{Deserialize, Serialize};
use sqlx::SqlitePool;
use std::path::PathBuf;

#[derive(sqlx::Type)]
#[sqlx(transparent)]
pub struct AppleCoreDateTime(f64);

impl From<AppleCoreDateTime> for chrono::DateTime<chrono::Utc> {
    fn from(value: AppleCoreDateTime) -> Self {
        chrono::NaiveDateTime::new(
            chrono::NaiveDate::from_ymd_opt(2001, 1, 1).expect("must succeed"),
            chrono::NaiveTime::from_hms_opt(0, 0, 0).expect("must succeed"),
        )
        .and_utc()
            + chrono::Duration::microseconds((value.0 * 1_000_000.0f64) as i64)
    }
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct Note {
    #[sqlx(try_from = "AppleCoreDateTime")]
    creation_date: chrono::DateTime<chrono::Utc>,
    subtitle: Option<String>,
    text: Option<String>,
    title: Option<String>,
    id: String,
}

#[derive(sqlx::FromRow)]
struct Tag {
    title: String,
}

pub struct BearDatabase {
    pool: SqlitePool,
}

impl BearDatabase {
    pub async fn new(db_path: PathBuf) -> Result<Self> {
        let database_url = format!("sqlite://{}?mode=ro", db_path.display());
        let pool = SqlitePool::connect(&database_url).await?;
        Ok(Self { pool })
    }

    pub async fn list_notes(&self, query: Option<&str>, tag: Option<&str>) -> Result<Vec<Note>> {
        let search_pattern: Option<String> = query.map(|query| format!("%{query}%"));
        let search_pattern_str = search_pattern.as_deref();

        Ok(sqlx::query_as!(
            Note,
            "SELECT
                creation_date AS 'creation_date!: AppleCoreDateTime',
                subtitle,
                text,
                title,
                id AS 'id!'
            FROM (
                SELECT
                    notes.ZCREATIONDATE AS creation_date,
                    notes.ZSUBTITLE AS subtitle,
                    notes.ZTEXT AS text,
                    notes.ZTITLE AS title,
                    notes.ZUNIQUEIDENTIFIER AS id,
                    (
                        CASE
                            WHEN ? IS NULL THEN 1.0
                            WHEN notes.ZTITLE = ? THEN 0.75
                            WHEN notes.ZTITLE LIKE ? THEN 0.5
                            WHEN notes.ZTEXT LIKE ? THEN 0.25
                            ELSE 0.0
                        END
                    ) AS text_search_score
                FROM 
                    ZSFNOTE notes
                LEFT JOIN 
                    Z_5TAGS note_tags ON notes.Z_PK = note_tags.Z_5NOTES
                INNER JOIN 
                    ZSFNOTETAG tags ON tags.Z_PK = note_tags.Z_13TAGS
                WHERE 
                    (? IS NULL OR tags.ZTITLE = ?)
                    AND text_search_score > 0
            )
            LIMIT
                10",
            search_pattern_str,
            query,
            search_pattern_str,
            search_pattern_str,
            tag,
            tag
        )
        .fetch_all(&self.pool)
        .await?)
    }

    pub async fn list_tags(&self) -> Result<Vec<String>> {
        Ok(sqlx::query_as!(
            Tag,
            "SELECT
                ZTITLE AS 'title!'
            FROM
                ZSFNOTETAG"
        )
        .fetch_all(&self.pool)
        .await?
        .into_iter()
        .map(|tag| tag.title)
        .collect())
    }
}
