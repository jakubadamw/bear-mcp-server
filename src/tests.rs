use std::path::PathBuf;

use crate::bear::BearDatabase;

#[rstest::rstest]
#[case(0, None, Some("first_tag"))]
#[case(1, None, Some("second_tag"))]
#[case(2, Some("Lorem"), None)]
#[case(3, Some("Test note with multiple tags"), None)]
#[case(4, Some("Test note with another tag"), Some("second_tag"))]
#[tokio::test]
async fn test_list_notes(
    #[case] index: usize,
    #[case] query: Option<&str>,
    #[case] tag: Option<&str>,
) {
    let path = PathBuf::from(env!("CARGO_MANIFEST_PATH"))
        .parent()
        .expect("must have a parent")
        .join("test_data")
        .join("database.sqlite");

    let bear_database = BearDatabase::new(path).await.expect("must succeed");
    let result = bear_database
        .list_notes(query, tag)
        .await
        .expect("must succeed");
    let snapshot_file_name = format!("test_list_notes_{index:02}");
    insta::assert_json_snapshot!(snapshot_file_name, result);
}
