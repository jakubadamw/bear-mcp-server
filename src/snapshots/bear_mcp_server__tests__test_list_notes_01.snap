---
source: src/tests.rs
expression: result
---
[
  {
    "creation_date": "2025-10-18T06:49:23.465508Z",
    "subtitle": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur cursus, felis ac ultrices luctus, leo mauris iaculis erat, ac mollis ante purus nec nibh. Vivamus eget tempor nunc. Duis id convallis",
    "text": "# Test note with multiple tags\n\n#first_tag\n#second_tag\n\nLorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur cursus, felis ac ultrices luctus, leo mauris iaculis erat, ac mollis ante purus nec nibh. Vivamus eget tempor nunc. Duis id convallis leo. Sed facilisis ornare dapibus. Mauris in tristique urna. Fusce vel diam arcu. \nPellentesque in sodales eros. Pellentesque fermentum ipsum eget elit imperdiet, sed molestie est aliquet. Fusce eget leo consectetur, feugiat magna at, consectetur neque. Nullam suscipit vehicula diam, sed consectetur dui mollis in. Sed eu dolor nisi. Interdum et malesuada fames ac ante ipsum primis in faucibus. Nulla placerat consequat ex, nec tincidunt elit consectetur ac. Etiam tempus, lectus ac suscipit suscipit, diam ipsum vulputate massa, vel accumsan leo felis in urna. Sed quis ultricies est, eu tempor erat.\n\nPraesent at justo et dui finibus rhoncus ut eleifend orci. Duis ac luctus massa. Sed gravida ante id neque fermentum lacinia et at nisl. Ut eu semper ante, eu commodo turpis. Integer sit amet urna id sem laoreet eleifend sit amet at mi. Nunc rhoncus mauris eros, in vulputate purus eleifend at. Vivamus ac auctor metus, at tempus turpis. Suspendisse eu vulputate risus. Vestibulum egestas, odio mollis sodales fermentum, dui leo eleifend dui, rhoncus ultrices quam felis posuere ex. Quisque fringilla diam pellentesque quam tincidunt scelerisque. Ut porta nibh eu tellus egestas imperdiet. In rhoncus vehicula nulla, aliquam pulvinar augue accumsan eu. Pellentesque sed gravida velit, vitae semper tortor.\nDonec ultrices risus eu augue luctus, id cursus mauris accumsan. Donec tempus mattis ultrices. Nam maximus vel ante sed molestie. Quisque eget libero eu justo venenatis hendrerit. Aenean molestie luctus lorem eu dapibus. Aenean eget laoreet nisi. Quisque semper, tellus ac tincidunt tempus, purus arcu pretium ex, porta malesuada metus neque a est. Suspendisse augue lorem, condimentum sit amet blandit in, sollicitudin sit amet mauris. Proin volutpat nibh orci, nec varius felis semper vel.\n\nMorbi quis velit sit amet arcu euismod lobortis. Sed bibendum leo id facilisis varius. Fusce nec mollis diam. Nullam finibus ut risus non fringilla. In non placerat mi. Duis ornare velit non lobortis vestibulum. Cras orci lacus, auctor et volutpat tempor, dignissim a lectus. Praesent auctor faucibus nisi ut sollicitudin. Vestibulum et est id leo rutrum commodo et vel lectus. Pellentesque vitae finibus neque.\n\nInteger sagittis ligula quis felis imperdiet hendrerit. Sed justo sem, maximus sed nunc in, tristique scelerisque enim. Interdum et malesuada fames ac ante ipsum primis in faucibus. Nullam pharetra massa a mi volutpat, a gravida mauris convallis. Aliquam non ullamcorper augue. Cras velit felis, tristique facilisis interdum ac, cursus sed urna. Sed eu lobortis purus. Pellentesque sed odio faucibus, viverra diam eget, porta ligula. Aenean condimentum tempus odio ac pharetra. Sed tempus massa sem, id tristique velit ultricies vitae.\n\nFusce congue condimentum condimentum. Ut quis sapien purus. Duis lacinia semper velit eget placerat. Curabitur sit amet eros tristique, sodales augue id, maximus nunc. Suspendisse aliquet libero turpis, eget ullamcorper odio ultricies et. Proin eleifend imperdiet dolor vitae euismod. Nulla nec velit massa. In ullamcorper urna quis sem maximus pellentesque. Phasellus vel magna eu mauris sollicitudin molestie et eu sem. Nulla dictum, turpis at dignissim vulputate, lectus risus gravida ligula, sit amet maximus neque neque vitae arcu. Donec egestas libero nec mi ultricies, quis ultrices diam faucibus. Morbi consectetur eleifend metus et egestas. Curabitur sed nibh in diam tincidunt cursus.\n",
    "title": "Test note with multiple tags",
    "id": "C94A52FA-6A48-41B4-981E-72B22B077438"
  }
]
